src/passes/PostAssemblyScript.cpp

```diff
@@ -0,0 +1,530 @@
+      *location = builder.makeNop();
+    } else {
+      assert((*location)->is<Nop>()); // already replaced
+    }
+  }
+
+  // Tests if a retain reaches an escape and thus is
+  // considered necessary.
+  bool testReachesEscape(LocalSet* retain, AliasGraph& graph) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      if (releases.find(localGet) != releases.end()) {
+        continue;
+      }
+      if (escapes.find(localGet) != escapes.end()) {
+        return true;
+      }
+      for (auto* localSet : graph.getInfluences[localGet]) {
+        if (testReachesEscape(localSet, graph)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  // Collects all reachable releases of a retain
+  void collectReleases(LocalSet* retain,
+                       AliasGraph& graph,
+                       std::unordered_set<Expression**>& found) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      auto foundRelease = releases.find(localGet);
+      if (foundRelease != releases.end()) {
+        found.insert(foundRelease->second);
+      } else {
+        for (auto* localSet : graph.getInfluences[localGet]) {
+          collectReleases(localSet, graph, found);
+        }
+      }
+    }
+  }
+
+  // Given a retain, gets the retained expression
+  static Expression* getRetainedExpression(LocalSet* retain) {
+    // assumes local.set(X, __retain(...))
+    auto* retainCall = retain->value->cast<Call>();
+    assert(retainCall->operands.size() == 1);
+    return retainCall->operands[0];
+  }
+
+  // Tests if a retained value originates at an allocation
+  // and thus is considered necessary.
+  bool testRetainsAllocation(Expression* retained, AliasGraph& graph) {
+    if (auto* call = retained->dynCast<Call>()) {
+      if (call->target == ALLOC || call->target == ALLOCARRAY) {
+        return true;
+      }
+    } else {
+      if (auto* localGet = retained->dynCast<LocalGet>()) {
+        for (auto* localSet : graph.getSetses[localGet]) {
+          if (localSet != nullptr &&
+              testRetainsAllocation(localSet->value, graph)) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  // Given a release location, gets the release
+  static LocalGet* getReleaseByLocation(Expression** releaseLocation) {
+    // assumes __release(local.get(X, ...))
+    auto* releaseCall = (*releaseLocation)->cast<Call>();
+    assert(releaseCall->operands.size() == 1);
+    return releaseCall->operands[0]->cast<LocalGet>();
+  }
+
+  // Tests if a release has balanced retains
+  bool testBalancedRetains(LocalGet* release,
+                           AliasGraph& graph,
+                           std::unordered_map<LocalGet*, bool>& cache) {
+    auto cached = cache.find(release);
+    if (cached != cache.end()) {
+      return cached->second;
+    }
+    for (auto* localSet : graph.getSetses[release]) {
+      if (localSet == nullptr) {
+        cache[release] = false;
```

Comment:
these can be `return cache[release] = false;`

Reply:
Heh, that's something I got used to on our end where doing this yields an unnecessary `load` of something just `store`d. Certainly not an issue for a C++ compiler, did change.